### ISSUE TITLE
int8-KV tensor-core flash-decode (`fa_split_gqa_mma_i8_kernel`) is only correct at HEAD_DIM=128 — the hd256 instantiations mix a 128-wide score/output stride with a HEAD_DIM-wide WMMA stride

### ISSUE THREAD

**Issue by boskodev790** [gittensor-ai-lab/sparkinfer#425]
## Summary

`fa_split_gqa_mma_i8_kernel<HEAD_DIM, GQA>` in
[`kernels/csrc/cuda/attention/flash_decode_split.cu`](kernels/csrc/cuda/attention/flash_decode_split.cu)
stores/loads its WMMA tiles with leading dimension **`HEAD_DIM`**, but the scalar softmax, the P′
quantization, and the output-accumulate epilogue index the same shared buffers with a hard-coded
stride of **128**. Those two strides coincide only when `HEAD_DIM == 128`, so the `<128, 8>`
instantiation is correct. The `<256, 8>` and `<256, 4>` instantiations — which **are** dispatched
for Qwen3.6 full-attention and Qwythos hd256 — read query rows ≥ 1 from the wrong offsets and only
ever compute the first 128 of the 256 output head dims. The result is a corrupted attention output,
not just a slower one.

This is likely a rework rather than a one-line fix, so I'm filing it as an issue for maintainer
review rather than a PR.

## Where it's dispatched

`launch_flash_decode_split` (same file), `head_dim == 256` branch:

- GQA-4 (Qwythos 16Q/4KV): line ~695 — `fa_split_gqa_mma_i8_kernel<256, GQA><<<..., i8_smem, ...>>>`
- GQA-8 (Qwen3.6 full-attn): line ~724 — `fa_split_gqa_mma_i8_kernel<256, GQA><<<..., i8_smem, ...>>>`

Both fire when:

```
head_dim == 256
&& int8_kv
&& mma_ok256  ==  (SPARKINFER_FAMMA != "0")   // default on
              && seqlen > 512
              && block_size == 16
              && chunk (= ceil(seqlen / n_splits)) >= 32
```

i.e. the long-context, int8-KV decode path — the same regime the 16k/32k benchmarks exercise.

## Shared-memory layout (for reference)

From the kernel prologue (line ~423):

```cpp
signed char* s_qi = ...;                 // [16][HEAD_DIM] quantized Q
signed char* s_pi = s_qi + 16*HEAD_DIM;  // [16][HEAD_DIM] quantized P'
float*       s_s  = ...(s_pi+16*HEAD_DIM);// [16][HEAD_DIM] scores / int32 mma scratch
float*       s_o  = s_s + 16*HEAD_DIM;    // [GQA][HEAD_DIM] running O
...
float*       s_ks = ...;                  // [128] group K scales   (128 == max KV tokens / group)
float*       s_vs = ...;                  // [128] group V scales
```

`s_s` and `s_o` are sized `16*HEAD_DIM` / `GQA*HEAD_DIM`, so the HEAD_DIM-strided WMMA stores stay
**in bounds** — this is a wrong-offset bug, not an out-of-bounds write.

## Root cause 1 — QK score store/read stride mismatch

The score matrix is `[16 query rows × (≤128) key tokens]`; 8 warps × 16 tokens = 128 columns, so its
row stride is the **key-token count (128)**, independent of HEAD_DIM.

- **Store (line ~485):** `store_matrix_sync(reinterpret_cast<int*>(s_s) + warp*16, cf, HEAD_DIM, mem_row_major);`
  → writes score(row `r`, keytoken `warp*16+c`) to `s_s[r*HEAD_DIM + warp*16 + c]`.
- **Read (line ~506):** `... (float)s_si[r * 128 + t] ...`  → reads `s_s[r*128 + t]`.

At HEAD_DIM=256 the store puts row `r` at stride 256 while the softmax reads it at stride 128, so for
every query row `r ≥ 1` the softmax reads a region the store never wrote for that row (e.g. row 1
reads `s_s[128..255]`, but row 1's scores were written to `s_s[256..383]`). Same 128-vs-HEAD_DIM
split at the P′ stash/quant (lines ~521, ~530).

## Root cause 2 — PV epilogue covers only 128 of 256 output dims, and re-mixes the stride

- Each warp writes 16 output head-dim columns: `vb = ... + warp*16` (line ~541), stored with ldm
  `HEAD_DIM` (line ~548). With 8 warps that is **128 columns**; for HEAD_DIM=256 the upper 128 head
  dims (128..255) are never computed. (Both `<256,8>` and `<256,4>` launch with 8 warps —
  `fa_mma_block_threads<256,4>::v == 256` threads, see line ~393.)
- The accumulate (line ~553): `for (i < GQA*128) s_o[i] += (float)((int*)s_s)[i] * s_ps[i>>7];`
  indexes `s_o` and `s_s` with stride 128 (`i>>7`), while the correction (line ~531) and the final
  write-out (lines ~561-562) index `s_o` with stride `HEAD_DIM`. At HEAD_DIM=256 those disagree, so
  the epilogue reads the PV result and writes `s_o` at the wrong rows on top of only-half-populated
  data.

## Why HEAD_DIM=128 is fine

Every `128` above equals `HEAD_DIM` when HEAD_DIM=128, and 8 warps × 16 = 128 covers the whole head,
so the `<128, 8>` path (Qwen3 hd128) is self-consistent. The bug is specific to the 256-wide
instantiations.

## Impact / open question

If the int8-KV hd256 MMA path is active during the scored long-context runs, its decode output is
numerically wrong even though it benchmarks fast. Worth confirming whether it slips past the
accuracy gate — e.g. if `accuracy.sh` runs at a context/`n_splits` where `mma_ok256` is false
(`seqlen ≤ 512`), or with bf16 KV (so the scalar/tile fallback `fa_split_gqa_kernel<256,...>` runs
instead), the gate wouldn't exercise this kernel while the 16k/32k **speed** numbers still would.
Setting `SPARKINFER_FAMMA=0` forces the scalar fallback and should be a clean A/B for correctness.

## Suggested direction (not a patch)

1. Store/read the QK scores at the key-token stride (128), not `HEAD_DIM` — i.e. pass `128` as the
   `store_matrix_sync` ldm at line ~485 (and correspondingly the PV store at ~548 needs the output
   laid out so the epilogue's stride matches).
2. Cover all `HEAD_DIM` output columns in the PV MMA — for HEAD_DIM=256 with 8 warps that means two
   column passes (warp handles `warp*16` and `warp*16 + 128`), or 16 warps.
3. Make the `s_o` accumulate/correct/write-out all use one consistent stride (`HEAD_DIM`).

## Repro sketch

Qwen3.6-35B-A3B (or Qwythos-9B) decode, int8 KV cache, seqlen > 512 (e.g. 4k/16k/32k),
`block_size == 16`, default `SPARKINFER_FAMMA`. Compare per-token logits / KL against the same run
with `SPARKINFER_FAMMA=0` (scalar fallback) or bf16 KV; a divergence that appears only in the
int8-KV hd256 long-context configuration would confirm it.

*(Filed from a source read; I don't have an RTX 5090 to run the A/B — flagging for maintainers who
can, since it touches a scored path and warrants verification before trusting hd256 int8-KV numbers.)*

**Comment by skyrocket2026** [comment]
Duplicate of #388 / closed #299 — same hd256 int8-MMA stride bug in `fa_split_gqa_mma_i8_kernel`. Tracked under #388 (regression re-land) with maintainer evidence; keeping one canonical issue.
